core/src/main/java/org/mskcc/cbio/portal/dao/DaoGenePanel.java

```diff
@@ -199,114 +197,137 @@ public static void updateGenePanel(GenePanel genePanel, Set<CanonicalGene> canon
         }
 
         try {
-        	updateGenePanelGeneList(internalId, canonicalGenes);
+            updateGenePanelGeneList(internalId, canonicalGenes);
         } finally {
             JdbcUtil.closeAll(DaoGenePanel.class, con, pstmt, rs);
         }
     }
 
-    
-    /*copied from addGenePanelGeneList above*/
-    private static void updateGenePanelGeneList(Integer internalId, Set<CanonicalGene> canonicalGenes) throws DaoException {
+    /**
+     * Gets a pair representing the sets of genes to add and to remove from a gene
+     * panel in the database.
+     * 
+     * @param incoming
+     * @param original
+     * @return
+     */
+    private static Object[] getAddRemove(Set<CanonicalGene> incoming, Set<CanonicalGene> original) {
+        // A -> incoming, B -> original
+        Set<CanonicalGene> toAdd = new HashSet<CanonicalGene>();
+        Set<CanonicalGene> toRemove = new HashSet<CanonicalGene>();
+        Set<CanonicalGene> AnB = intersectionTwoLists(new ArrayList<>(incoming), new ArrayList<>(original));
+        Set<CanonicalGene> AuB = unionTwoLists(new ArrayList<>(incoming), new ArrayList<>(original));
+
+        // 1. Is A == B?
+        if (AnB.equals(AuB)) {
+            ;
+        }
+
+        // 2. Is A ^ B = nil?
+        else if (AnB.isEmpty()) {
+            // Remove B
+            toRemove = original;
+            // Add A
+            toAdd = incoming;
+        }
+
+        // 3. Is AnB == B?
+        else if (AnB.equals(original)) {
+            // Add A - B
+            toAdd = diffTwoLists(new ArrayList<>(incoming), new ArrayList<>(original));
+        }
+
+        // 4. Is AnB == A?
+        else if (AnB.equals(incoming)) {
+            // Remove B - A
+            toRemove = diffTwoLists(new ArrayList<>(original), new ArrayList<>(incoming));
+        }
+
+        // 5. Intersect but neither are supersets of one other
+        else {
+            // Add A - AnB
+            toAdd = diffTwoLists(new ArrayList<>(incoming), new ArrayList<>(AnB));
+            // Remove B - AnB
+            toRemove = diffTwoLists(new ArrayList<>(original), new ArrayList<>(AnB));
+        }
+
+        return new Object[] { toAdd, toRemove };
```

Comment:
not sure if I should've used a Pair instead, but the type casting messages to Set<CanonicalGene> don't seem to crash anything